Assorted fixes for runtime metadata mangling and demangling

rjmccall opened the pull request:
Primarily, teach the compiler to not use the runtime demangler to produce `@MainActor (...) -> sending T` and similar type metadata unless the deployment includes the bug fix in Swift 6.1. But I've also fixed the metadata-to-demangling-tree walk to preserve the `@isolated(any)` bit, and I've removed some code that was duplicating the information in `RuntimeVersions.def` and implicitly making aggressive assumptions about future OS releases.

rjmccall committed:
Fix remangling isolated(any) function types from type metadata.

rjmccall committed:
Generate getSwiftRuntimeCompatibilityVersionForTarget from
RuntimeVersions.def.

This really *ought* to be NFC, but alas, there are two behavior changes.
The first is that arbitrary future versions (short of the special future
version of 99.99) used to be considered to be free of all deployment
issues, but are now just considered to offer the Swift release that's
attached to the last known release of the target platform. The old behavior
was fine in a sense: deployment issues require updating the compiler anyway,
so if you haven't taught the compiler that iOS 31 provides Swift 8.3, you
probably also haven't taught it about any problems that are fixed in
Swift 8.3. The only problem with this logic is that we often implement
compiler fixes for these deployment issues before that first OS actually
ships (and thus before we've settled on a release number), and so we end
up with a race where we start considering new releases to fix the issue
even without updating the release mapping. The new approach fixes this:
we say that the bug will be fixed in Swift 8.3, and for the time being,
there are simply no platforms that provide that fix yet. Later, we update
the mapping to say that iOS 31 provides it, and compilations targeting that
release will be able to take advantage.

The other issue is that we actually treated "future" releases of macOS 10.x
(starting with 10.16, which ended up being 11.0) the same way, instead of
treating them logically as intermediate between 10.15 and 11.0.

My initial attempt at this generation used a constexpr array of a struct
with a std::initializer_list of platform releases, exactly mirroring the
structure of the .def file and requiring the compiler to parse out the
bits relevant to the target dynamically. The approach I ended up with is
much better, but I was actually forced into it because MSVC miscompiled
those global "temporary" arrays --- they ended up uninitialized.

rjmccall committed:
Fix a pair of IRGen bugs around materializing function type metadata
with sending results:

- The sending result mangling was added in the 6.0 runtime, so demangling
  cannot be used to produce this metadata when targeting an earlier
  runtime.

- The combination of a sending result with isolation requires the 6.1
  runtime to successfully demangle, due to a bug in the 6.0 demangler.

rjmccall commented:
@swift-ci Please test

tshortli commented on lib/Basic/Platform.cpp:
```diff
@@ -431,147 +431,139 @@ llvm::Triple swift::getUnversionedTriple(const llvm::Triple &triple) {
+
+enum class Platform {
+  macOS,
+  iOS,
```
Is tvOS intentionally excluded because it never diverges from iOS?

rjmccall replied:
Yeah, that assumption is already pervasive in `RuntimeVersions.def`.

rjmccall committed:
NFC: use visionOS instead of xrOS in the Swift codebase. We can't do
much about the LLVM triple, but we can do our best on the Swift side.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

AnthonyLatsis commented:
@swift-ci Please test

TheHarcker commented on lib/Basic/Platform.cpp:
```diff
@@ -431,147 +431,139 @@ llvm::Triple swift::getUnversionedTriple(const llvm::Triple &triple) {
-  }
-  else if (Triple.isXROS()) {
+
+    // tvOS releases predate the stable ABI, so use Swift 5.0 as our base.
```
Shouldn't this be watchOS?
```suggestion
    // watchOS releases predate the stable ABI, so use Swift 5.0 as our base.
```

rjmccall replied:
Yes, absolutely. Thanks for catching the slip-up.

rjmccall commented:
@swift-ci Please test

rjmccall committed:
Fix test to use macOS 12.0 as the first OS to not require autolinking
instead of the notional 10.24, which the old logic in Platform.cpp
accidentally treated as a future release with no runtime restrictions.

rjmccall commented:
@swift-ci Please test

CodaFi commented on lib/Basic/Platform.cpp:
```diff
@@ -431,147 +431,139 @@ llvm::Triple swift::getUnversionedTriple(const llvm::Triple &triple) {
+      // this must be the highest supported Swift release.
+      if (provider.platform == targetPlatform &&
+          provider.version <= targetVersion) {
+        return MAX(swiftRelease.version, minimumRelease);
```
Can we use `std::max` instead?

rjmccall replied:
The existing code wasn't, and that seemed deliberate. I didn't look into it.

CodaFi replied:
Perhaps it was because this code was written before we had C++14 as the default language mode.

CodaFi replied:
Alright, I'll get it https://github.com/swiftlang/swift/pull/78819

rjmccall commented:
@swift-ci Please test

rjmccall committed:
Add a header for various common variadic-templates metaprogramming tasks.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test macOS

rjmccall commented:
@swift-ci Please test Linux
